File: PHP/Tree.php
<?php

require "./TreeNode.php";

class Tree
{
  private ?TreeNode $root;

  public function __construct()
  {
    $this->root = null;
  }

  public function addValue(int $value)
  {
    $newNode = new TreeNode($value);
    $this->root = TreeNode::add($this->root, $newNode);
  }

  public function searchValue($value)
  {
    $result = TreeNode::search($this->root, $value);

    if ($result) {
      echo "Valor encontrado: ". $result->value. "\n";
    }
    
    echo "Valor não encontrado\n";
  }

  public function deleteValue($value)
  {
    $this->root = TreeNode::delete($this->root, $value);
  }

  public function print()
  {
    echo "Árvore\n";
    echo TreeNode::print($this->root). "\n";
    echo "\n";
  }

  public function numberOfNodesInTree()
  {
    return TreeNode::count($this->root);
  }
}

File: Go/tree/tree.go
package tree

type TreeNode struct {
	value int
	dir *TreeNode
	esq *TreeNode
}

func AddNode(root *TreeNode, newNode *TreeNode) *TreeNode {
	if root == nil {
		return newNode
	}

	if newNode.value < root.value {
		root.esq = AddNode(root.esq, newNode)
	}
	else {
		root.dir = AddNode(root.dir, newNode)
	}

	return root
}

func AddNode(root *TreeNode, newNode *TreeNode) *TreeNode {
	if root == nil {
		return newNode
	}

	if newNode.value < root.value {
		root.esq = AddNode(root.esq, newNode)
	}
	else {
		root.dir = AddNode(root.dir, newNode)
	}
	return root
}

func SearchValue(root *TreeNode, value int) *TreeNode {
	switch {
	case root == nil:
		return nil
	case root.value == value:
		return root
	case root.value > value:
		return SearchValue(root.esq, value)
	default:
		return SearchValue(root.dir, value)
	}
}

func CountNodes(root *TreeNode) int {
	if root == nil { return 0 }

	return CountNodes(root.esq) + 1 + CountNodes(root.dir)
}